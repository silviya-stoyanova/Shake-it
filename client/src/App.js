import React, { Component } from 'react';
import { BrowserRouter } from 'react-router-dom'
import { ToastContainer } from 'react-toastify'
import 'react-toastify/dist/ReactToastify.css';
import './static/css/site.css'
import './static/css/scrollbar.css'

import sessionManager from './components/utilities/session-util'
import observer from './components/utilities/observer'

import MyRoutes from './components/routes/router'
import Header from './components/common/header'
import Footer from './components/common/footer'

const defaultValue = { isLogged: false }
const { Consumer: UserInfoConsumer, Provider: UserInfoProvider } = React.createContext(defaultValue)

class App extends Component {
  constructor(props) {
    super(props)
    this.state = {
      isLogged: sessionManager.isUserLogged(),
      username: sessionManager.getUsername(),
      role: sessionManager.getUserRole()
    }

    observer.subscribe('userLogin', this.onUserLogin)
    observer.subscribe('userLogout', this.onUserLogout)
  }

  onUserLogin = () => {
    this.setState(prevState => ({ // asyncronous
      isLogged: true,
      username: sessionManager.getUsername(),
      role: sessionManager.getUserRole()
    }))
  }

  onUserLogout = () => {
    this.setState(prevState => ({
      isLogged: false,
      username: '',
      role: []
    }))
  }

  render() {
    const { isLogged, username, role } = this.state

    // console.log(this.props.navigation)

    return (
      <UserInfoProvider value={{ isLogged, username, role }}>
        <BrowserRouter>
          <ToastContainer />
          <Header />

          {
            // <div className="main-container">
          }
          <MyRoutes />
          {
            // </div>
          }

          <Footer />
        </BrowserRouter>
      </UserInfoProvider>
    )
  }

  componentDidMount() {
    sessionManager.isUserLogged()
      ? this.setState({ isLogged: true })
      : this.setState({ isLogged: false })
  }
}

export { UserInfoConsumer }
export default App