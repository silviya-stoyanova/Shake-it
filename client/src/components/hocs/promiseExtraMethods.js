import requester from '../../utilities/requests-util'
import observer from '../../utilities/observer'
import sessionManager from '../../utilities/session-util'

const promiseExtraMethods = {
    user: {
        onLoginSuccess(res) {
            sessionManager.saveSession(res.authtoken, res.username, res.role)
            observer.trigger('userLogin')
        },
        onRegisterSuccess(res, data) {
            let nextPromise = requester.login(data)
            this.handleFetchPromise(nextPromise, this.onLoginSuccess, this.onLoginFail, data) // { username: data.username, password: data.password }
        },
        onLoginFail(err) {
            this.setState(prevState => ({
                data: {
                    ...prevState.data,
                    userClass: 'correct',
                    passClass: 'error',
                    repeatPassClass: 'error'
                }
            }))
        },
        onRegisterFail(err) {
            this.setState(prevState => ({
                data: {
                    ...prevState.data,
                    userClass: 'error',
                    passClass: 'correct',
                    repeatPassClass: 'correct'
                }
            }))
        }
    },

    product: {
        onProductPromiseSuccess(err) {
            return err // do nothing
        },

        onProductPromiseFail(err) {
            if (!err.titleIsTaken) {
                this.props.history.push('/')
            }
        }
    }
}

export default promiseExtraMethods