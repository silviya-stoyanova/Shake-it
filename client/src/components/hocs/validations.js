import { toast } from 'react-toastify'

const userValidations = {
    validateUserData(formType, username, password, repeatPassword, updateState) {
        let [userClass, passClass, repeatPassClass] = ['error', 'error', 'error']
        let result = {
            username: 'invalid',
            password: 'invalid',
            repeatPassword: 'invalid'
        }

        if (username && username.trim().length >= 5) {
            userClass = 'correct'
            result.username = 'valid'
        }
        if (password && password.length >= 8) {
            passClass = 'correct'
            result.password = 'valid'
        }

        if (repeatPassword) {
            if (repeatPassword && repeatPassword.length >= 8) {
                repeatPassClass = 'correct'
                result.repeatPassword = 'valid'
            }

            if (password !== repeatPassword) {
                [passClass, repeatPassClass] = ['error', 'error']
                result.password = 'invalid'
                result.repeatPassword = 'invalid'
            }
        }

        updateState && this.setState(prevState => ({
            userInfo: {
                ...prevState.userInfo,
                userClass,
                passClass,
                repeatPassClass
            }
        }))

        return result
    },

    validateUserOnSubmit(formType, username, password, repeatPassword) {
        const isInputValid = this.validateUserData(formType, username, password, repeatPassword, false)
        let isValid = false
        let errorMsg = ''

        if (formType === 'login') {
            if (isInputValid.username === 'invalid' || isInputValid.password === 'invalid') {
                errorMsg = 'Either your username or password is invalid.'
            }
        } else if (formType === 'register') {
            if (isInputValid.password === 'invalid' || isInputValid.repeatPassword === 'invalid') {
                errorMsg = 'Both passwords must consist of at least 8 characters and they both must match!'
            }
            if (isInputValid.username === 'invalid') {
                errorMsg = 'Please enter a valid username!'
            }
        }
        if (errorMsg) {
            toast.info(errorMsg, {
                className: 'error-toast'
            })
            return isValid
        }

        isValid = true
        return isValid
    }
}

const productValidations = {
    validateProductData(formType, title, description, image, price, updateState) {
        let [titleClass, descriptionClass, priceClass] = ['error', 'error', 'error', 'error']
        let result = {
            title: 'invalid',
            description: 'invalid',
            image: 'invalid',
            price: 'invalid',
        }

        if (title && title.length > 2 && title.length <= 50) {
            titleClass = 'correct'
            result.title = 'valid'
        }
        if (description && description.length > 9 && description.length <= 250) {
            descriptionClass = 'correct'
            result.description = 'valid'
        }

        // validate only when creating a new product
        if (formType === 'create' && image) {
            result.image = 'valid'
        }
        if (price && Number(price) > 0) {
            priceClass = 'correct'
            result.price = 'valid'
        }

        updateState && this.setState(prevState => ({
            productInfo: {
                ...prevState.productInfo,
                titleClass,
                descriptionClass,
                priceClass
            }
        }))
        return result
    },

    validateProductOnSubmit(formType, title, description, image, price) {
        const updateState = false
        const isInputValid = this.validateProductData(formType, title, description, image, price, updateState)
        let isValid = false
        let errorMsg = ''

        if (isInputValid.price === 'invalid') {
            errorMsg = 'Please enter a valid price.'
        }
        // when adding a product
        if (formType === 'create' && isInputValid.image === 'invalid') {
            errorMsg = 'Please provide an image.'
        }
        if (isInputValid.description === 'invalid') {
            errorMsg = 'The description must be at least 10 symbols long.'
        }
        if (isInputValid.title === 'invalid') {
            errorMsg = 'The title length must be between 3 and 50 symbols including.'
        }
        if (errorMsg) {
            toast.info(errorMsg, {
                className: 'error-toast',
            })
            return isValid
        }

        isValid = true
        return isValid
    }
}

export { userValidations }
export { productValidations }