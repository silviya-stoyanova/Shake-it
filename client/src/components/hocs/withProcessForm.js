import React, { Component } from 'react'
import { toast } from 'react-toastify'
import requester from '../../utilities/requests-util'
import sessionManager from '../../utilities/session-util'

const withProcessForm = (Form, formType, validations, initialData, requestType, promiseExtraMethods, testServiceOnMount, testServiceOnSubmit) => {
    return class extends Component {
        constructor(props) {
            super(props)
            this.state = {
                data: initialData
            }

            this.validateData = validations ? validations.validateData.bind(this) : null
            this.validateOnSubmit = validations ? validations.validateOnSubmit.bind(this) : null

            this.promiseSuccess = promiseExtraMethods ? promiseExtraMethods.success.bind(this) : null
            this.promiseFail = promiseExtraMethods ? promiseExtraMethods.fail.bind(this) : null

            this.handleInputChange = this.handleInputChange.bind(this)
            this.handleFetchPromise = this.handleFetchPromise.bind(this)
            this.handleFormSubmit = this.handleFormSubmit.bind(this)
        }

        // static defaultProps = {
        //     service: requester.getProductInfo
        // }

        async handleInputChange({ target }) {
            console.log(target.files ? target.files[0] : null)

            await this.setState(prevState => ({
                data: {
                    ...prevState.data,
                    [target.name]: target.files ? target.files : target.value, // .trim()
                    uploadedImg: target.files ? URL.createObjectURL(target.files[0]) : prevState.data.uploadedImg
                    // uploadedImg: target.files ? target.files[0] : prevState.data.uploadedImg
                }
            }))

            const updateState = true
            const { data } = this.state

            // console.log(data.uploadedImg)

            this.validateData(formType, data, updateState)
        }

        handleFetchPromise(promise, onSuccessFunc, onFailFunc, data = '') {
            promise.then(res => {
                if (!res.ok) {
                    return Promise.reject(res)
                }
                return res.json()

            }).then(res => {
                toast.info(res.success, {
                    className: 'success-toast'
                })

                // data.redirectOnSuccess &&
                this.props.history.push('/')

                onSuccessFunc && onSuccessFunc(res, data)

            }).catch(err => {
                err.json().then(error => {
                    toast.info(error.message, {
                        className: 'error-toast'
                    })

                    onFailFunc && onFailFunc(error)
                })
            })
        }

        async handleFormSubmit(event) {
            // console.log(event.preventDefault())
            
            event.preventDefault()
            const { data } = this.state
            const jwtToken = sessionManager.getUserInfo().authtoken
            let isValid = this.validateOnSubmit(formType, data)

            if (!isValid) {
                return
            }

            if (testServiceOnSubmit) {
                testServiceOnSubmit()
                    // to see the message inside the response, uncomment the next two commented out lines
                    // .then(res => res.json())
                    .then(res => {
                        // console.log(res.success)
                        this.props.history.push('/')
                    })
                    .catch(err => {
                        return err
                    })

            } else {
                const promise = requester[requestType](data, jwtToken)
                this.handleFetchPromise(promise, this.promiseSuccess, this.promiseFail, data)
            }
        }

        render() {
            return <Form {...this.state} handleInputChange={this.handleInputChange} handleFormSubmit={this.handleFormSubmit} />
        }

        componentDidMount() {
            if (formType === 'create' || formType === 'login' || formType === 'register') {
                return
            }
            // let promise = requester.getProductInfo(this.state.productInfo._id)
            // this.handleFetchPromise(promise, null, null, { redirectOnFail: true })
            // requester.getProductInfo(this.state.data._id)

            const serviceReq = testServiceOnMount || requester.getProductInfo  // new

            const productId = this.props.match                      // new
                ? this.props.match.params.productId
                : null
            // const productId = this.props.match.params.productId

            // requester.getProductInfo(productId)                  // prev
            serviceReq && serviceReq(productId)                     // new
                .then(res => {
                    if (!res.ok) {
                        return Promise.reject(res)
                    }
                    return res.json()
                })
                .then(async res => {
                    await this.setState({
                        data: res
                    })
                })
                .catch(error => {
                    this.props.history.push('/')

                    error.json().then(err => {
                        toast.info(err.message, {
                            className: 'error-toast'
                        })
                    })
                })
        }
    }
}

export default withProcessForm