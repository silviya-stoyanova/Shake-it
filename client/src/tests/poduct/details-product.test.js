import React from 'react'
import { mount } from 'enzyme'
import { UserInfoProvider } from '../../App'
import withProcessForm from '../../components/hocs/withProcessForm'
import ProductDetails from '../../components/product/details-product'

import { productValidations } from '../../components/hocs/validations'
import promiseExtraMethods from '../../components/hocs/promiseExtraMethods'

// https://dev.to/papaponmx/unit-testing-hoc-connected-components-2d8h
const mockProductFetching = () => {
    const sampleData = JSON.stringify({
        _id: '1',
        title: 'Test',
        description: 'This is a test for the React component ProductDetails',
        image: 'no-img',
        price: '1',
        likes: '100'
    })

    // wrap the data in a new Response object to be able to handle it correctly in withProcessForm.js
    return Promise.resolve(new Response(sampleData))
}

const initialData = {
    // _id: props.match.params.productId,
    title: '',
    description: '',
    image: '',
    price: '',
}

const extraMethods = {
    success: promiseExtraMethods.product().onProductPromiseSuccess,
    fail: promiseExtraMethods.product().onProductPromiseFail,
}

describe('tests for ProductDetails component', () => {
    it('should render product details without any buttons below it, when a user is not logged in', () => {
        const productPromise = mockProductFetching()

        const TestHoc = withProcessForm(<ProductDetails />, 'details', productValidations, initialData, null, extraMethods, mockProductFetching)

        const wrapper = mount(
            <UserInfoProvider>
                {TestHoc}
            </UserInfoProvider>
        )
        console.log(wrapper)

        // const wrapper = mount(
        //     withProcessForm(
        //         <UserInfoProvider>
        //             <ProductDetails />
        //         </UserInfoProvider>,
        //         'details', productValidations, initialData, null, extraMethods, mockProductFetching)
        // )

        // const wrapper = mount(
        //     <withProcessForm Form={
        //         <UserInfoProvider>
        //             <ProductDetails />
        //         </UserInfoProvider>}

        //         formType={'details'} validations={productValidations} initialData={initialData} requestType={null} promiseExtraMethods={extraMethods} service={mockProductFetching}
        //     />
        // )

        productPromise
            .then(res => {
                if (!res.ok) {
                    return Promise.reject(res)
                }
                return res.json()
            })
            .then(res => {
                // do stuff here
                wrapper.update()
            })


        expect(wrapper.debug()).toMatchSnapshot()
        // expect(wrapper.find('div.product-title').length).toEqual(1)
    })






})
