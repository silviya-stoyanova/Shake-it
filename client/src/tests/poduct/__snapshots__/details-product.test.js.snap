// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for ProductDetails component should render loading circle gif, when The [Admin] accesses [non-existing] product's details 1`] = `"<div class=\\"content-wrapper\\"><div class=\\"product-details-wrapper\\"><img src=\\"loading-circle.gif\\" alt=\\"loading-img\\" class=\\"product-img\\"></div></div>"`;

exports[`tests for ProductDetails component should render loading circle gif, when a [guest] accesses [non-existing] product's details 1`] = `"<div class=\\"content-wrapper\\"><div class=\\"product-details-wrapper\\"><img src=\\"loading-circle.gif\\" alt=\\"loading-img\\" class=\\"product-img\\"></div></div>"`;

exports[`tests for ProductDetails component should render loading circle gif, when a [user] accesses [non-existing] product's details 1`] = `"<div class=\\"content-wrapper\\"><div class=\\"product-details-wrapper\\"><img src=\\"loading-circle.gif\\" alt=\\"loading-img\\" class=\\"product-img\\"></div></div>"`;

exports[`tests for ProductDetails component should render product details with all links for [like][edit][delete][add to cart] below it, when The [Admin] accesses [existing] product's details 1`] = `"<div class=\\"content-wrapper\\"><div class=\\"product-details-wrapper\\"><div class=\\"product-title\\">Test</div><img src=\\"data:image/png;base64, no-img\\" alt=\\"Test\\" class=\\"product-img\\"><div class=\\"product-description\\">This is a test for the React component ProductDetails</div><div class=\\"product-actions\\"><span class=\\"product-price-details\\">Price: 1<span class=\\"price-sign\\">$</span></span><span class=\\"product-likes\\">4 ♥</span><a class=\\"product-actions-btn\\" href=\\"/product/like/1\\">like</a><a class=\\"product-actions-btn\\" href=\\"/product/edit/1\\">edit</a><a class=\\"product-actions-btn\\" href=\\"/product/delete/1\\">delete</a><a class=\\"product-actions-btn add-to-cart-btn\\" href=\\"/cart/add/1\\">add to cart</a></div></div></div>"`;

exports[`tests for ProductDetails component should render product details with links for [like] and [add to cart] below it, when a [user] accesses [existing] product's details 1`] = `"<div class=\\"content-wrapper\\"><div class=\\"product-details-wrapper\\"><div class=\\"product-title\\">Test</div><img src=\\"data:image/png;base64, no-img\\" alt=\\"Test\\" class=\\"product-img\\"><div class=\\"product-description\\">This is a test for the React component ProductDetails</div><div class=\\"product-actions\\"><span class=\\"product-price-details\\">Price: 1<span class=\\"price-sign\\">$</span></span><span class=\\"product-likes\\">4 ♥</span><a class=\\"product-actions-btn\\" href=\\"/product/like/1\\">like</a><a class=\\"product-actions-btn add-to-cart-btn\\" href=\\"/cart/add/1\\">add to cart</a></div></div></div>"`;

exports[`tests for ProductDetails component should render product details without any links below it, when a [guest] accesses [existing] product's details 1`] = `"<div class=\\"content-wrapper\\"><div class=\\"product-details-wrapper\\"><div class=\\"product-title\\">Test</div><img src=\\"data:image/png;base64, no-img\\" alt=\\"Test\\" class=\\"product-img\\"><div class=\\"product-description\\">This is a test for the React component ProductDetails</div><div class=\\"product-actions\\"><span class=\\"product-price-details\\">Price: 1<span class=\\"price-sign\\">$</span></span><span class=\\"product-likes\\">4 ♥</span></div></div></div>"`;
